### PR TITLE
Shelly

### PR DESCRIPTION
**Pull request by gyavolo**
This pull request covers several commits, which add the following improvements and fixes:

 - New feature: Add SPI frequency macro in mos.yml (MGOS_ADE7953_SPI_FREQ: 1000000).

 - Fix: Change SPI mode to 0. ADE7953 works with this mode only.

 - New feature: Add configurable PGA gain values for the voltage and current PGA registers (PGA_V, PGA_IA, PGA_IB). See the README.md for the new struct mgos_ade7953_config members.

 - New feature: Add power factor measurment function mgos_ade7953_get_pf.

 - Improvement (BREAKING CHANGE!): Move struct mgos_ade7953_config definition to mos.yml. Create abstract config schema, which makes possible to use config parameters directly from the device mgos_config. Due to the limitations for defining float type, arrays and the structure name, slight changes to struct mgos_ade7953_config are necessary (see the example in README.md ), The name is changed to mgos_config_ade7953.

### CODE REVIEW

**Comment by rojer** [comment]
float -> double change is more significant than it might seem, as floats are natively supported by some hardware platforms (ESP32, STM32) but none of the platforms support hardware double-precision arithmetic.

in fact, i wouldn't mind adding "f" type to the mos config generator. if you could put up a PR that does it, that'd be great.

File: include/mgos_ade7953.h [hunk 1]
```diff
@@ -92,6 +92,10 @@ bool mgos_ade7953_get_apower(struct mgos_ade7953 *dev, int channel, float *watts
 // If reset is true, resets the accumulator.
 bool mgos_ade7953_get_aenergy(struct mgos_ade7953 *dev, int channel, bool reset, float *wh);
 
+// Write the measured power factor to the *pf pointer.
```
Comment (pimvanpelt) [inline include/mgos_ade7953.h:71]: Can you specify the unit of pf? I suppose something like W/VA ?

from the fluke article on this -- Power factor (PF) is the ratio of working power, measured in kilowatts (kW), to apparent power, measured in kilovolt amperes (kVA)

File: README.md [hunk 2]
```diff
@@ -48,15 +48,20 @@ enum mgos_app_init_result mgos_app_init(void) {
   struct mgos_ade7953 *ade;
 
   // These constants are specific to the specific application circuit.
-  const struct mgos_ade7953_config ade_cfg = {
+  const struct mgos_config_ade7953 ade_cfg = {
```
Comment (pimvanpelt) [inline README.md:51]: OK for breaking change

**Review by pimvanpelt — CHANGES_REQUESTED** [review]
Changes look great, thank you for the contribution.
I didn't test the SPI settings (I think at all, actually), I assume you can validate that they work in shelly devices at least?

Just one small change, a documentation nit - I'm a sucker for units in casted return valies (in this case of the powerfactor function). Can you take a look? I'm happy to approve.

**Comment by rojer** [comment]
i added float config value type - https://github.com/cesanta/mongoose-os/commit/9a1170d7413b4fd39d2b68614bc6ce07d7a15775

File: include/mgos_ade7953.h (hunk 1 above)
Comment (gyavolo) [reply include/mgos_ade7953.h:71]: AFAIK, the power factor is a dimensionless number in the range [-1,1]. I don't think W/VA is appropriate.

**Comment by gyavolo** [comment]
> float -> double change is more significant than it might seem, as floats are natively supported by some hardware platforms (ESP32, STM32) but none of the platforms support hardware double-precision arithmetic.
> 
> in fact, i wouldn't mind adding "f" type to the mos config generator. if you could put up a PR that does it, that'd be great.

I noticed that you have added "f" type already, so amend-ed the last commit to use float.

**Comment by rojer** [comment]
LGTM

**Pull request merged**